sycl/source/detail/scheduler/graph_builder.cpp

```diff
@@ -533,7 +533,8 @@ Scheduler::GraphBuilder::addHostAccessor(Requirement *Req,
 
   if (sameCtx(HostAllocaCmd->getQueue()->getContextImplPtr(),
               Record->MCurContext)) {
-    if (!isAccessModeAllowed(Req->MAccessMode, Record->MHostAccess))
+    if (!isAccessModeAllowed(Req->MAccessMode, Record->MHostAccess) &&
+        !Req->MIsSubBuffer)
```

Comment:
The `remapMemoryObject()` call that this guards asserts that both the HostAllocaCmd and its LinkedAllocaCmd exist, and then uses them extensively.  But if this is a SubBuffer, there is no LinkedAllocaCmd. That's on the parent.  So making this guard reject sub-buffers completely is correct. 

I thought that the parent of the sub-buffer would be remapped, as part of visiting the graph nodes, but maybe I'm mistaken. Maybe the parent memory object should be remapped in that case, as you suggest?   Can you think of a quick way of testing this?

Reply:
> Maybe the parent memory object should be remapped in that case, as you suggest? Can you think of a quick way of testing this?
Yeah, I think so. As for testing it, I'm not sure if there's a way to reliably trigger an error by using the wrong access mode, it would probably be device dependent. Inspecting the PI trace logs to see if the remapping was triggered is probably the best way to check this quickly.

Reply:
@sergey-semenov  - I finally returned to this.  The parent buffer gets the Map/Unmap operations correctly, the sub-buffer doesn't have to do anything.  I have update the test to confirm this. This should be ready now.

Reply:
> The parent buffer gets the Map/Unmap operations correctly

I don't think it does. Here are the map API calls that are made in your test.
```
---> piEnqueueMemBufferMap(
  <unknown> : 0x1ddde98
  pi_mem : 0x1c9b468
  <unknown> : 0
  <unknown> : 1
  <unknown> : 0
  <unknown> : 200
  <unknown> : 1
  const pi_event * : 0x1de6f40[ 0x1de2258 ... ]
  pi_event * : 0x1de5e38[ 0 ... ]
  <unknown> : 0x7ffef70fcbf8
) --->  pi_result : PI_SUCCESS
  [out]pi_event * : 0x1de5e38[ 0x1de7088 ... ]
  [out]<unknown> ** : 0x7ffef70fcbf8[ 0x1de1e80 ... ]

---> piEnqueueMemBufferMap(
  <unknown> : 0x1ddde98
  pi_mem : 0x1de3478
  <unknown> : 0
  <unknown> : 3
  <unknown> : 0
  <unknown> : 200
  <unknown> : 1
  const pi_event * : 0x1de6f40[ 0x1de4058 ... ]
  pi_event * : 0x1de6dd8[ 0 ... ]
  <unknown> : 0x7ffef70fcbf8
) --->  pi_result : PI_SUCCESS
  [out]pi_event * : 0x1de6dd8[ 0x1de7508 ... ]
  [out]<unknown> ** : 0x7ffef70fcbf8[ 0x1de3b80 ... ]

```
The first one is made for the parent buffer, the second one - for `buf_y`. That first "1" stands for read only mode. The parent buffer is never remapped in this test, even though its buffer is later accessed with the `read_write` mode.

Reply:
@sergey-semenov  - I'm not sure if that analysis is correct. The buffer is mapped back so that the host operations will be successful. The read/write flag of the later host operation isn't relative.  "read" is the correct requirement when mapping BOTH buffers.

I can confirm that the values that are finally seen in both buffers after the host tasks are done is correct. If I add code that confirms that to the test would that  address your concerns?

Reply:
>The buffer is mapped back so that the host operations that come later will be successful. The read/write flag of the later host operation isn't relative. "read" is the correct requirement when mapping BOTH buffers.

"read_only" can't be correct for the parent buffer, since we perform write operations on the mapped memory after that by creating a read_write accessor to its subbuffer on L49, which is exactly what "read_only" forbids.

> I can confirm that the values that are finally seen in both buffers after the host tasks are done is correct. If I add code that confirms that to the test would that address your concerns?

Not really, at least not without expanding this test to include some operations on device after those host tasks. I'd expect everything to fall apart once we do unmap and work with those written values on the device. Even then though, it's undefined behavior, so whether it actually fails or not will probably be device dependent. I'd be very surprised if this works in its current state on a discrete GPU though.

Reply:
@sergey-semenov you are right and I was mistaken. I have updated the PR to submit the parentalloca and updated the tests to check for the read_write mode in the map as appropriate.